### PR TITLE
Automatic air raids

### PR DESCRIPTION
**Pull request by DanVanAtta**
Builds on top of: https://github.com/triplea-game/triplea/pull/1426

Included:
- Bombing and air raids are "auto-selected"
  - This reduces clicks in the game flow, these battles have to be fought
  - Game rules would have these battles fought first, then amphib. We are for now forcing that these battles be fought first. See #1426 for more discussion on purpose

Test notes:
@simon33-2  - quite right to point out that possible to test is not the same as easy to test. There are a number of scaffolding changes included just to make testing easier to write. 

Enhancements on #1426 
- added checks for SBR in addition to Air raids, and will fight those first as well
- added unit test + scaffolding

Now that we auto-fight SBR's, a few integration tests broke. I removed rather than fixing them since they can no longer test anything. They were dependent on battles still existing after calling BattleDelegate.start(), which is now no longer the case. With the unit test in place we have test coverage, so we can remove the integration tests.

### CODE REVIEW

**Comment by simon33-2** [comment]
One quick comment is that to be rules compliant all the air raids should be fought before all the bombing raids. This is because the intercept decisions are supposed to be made before any combat is performed, at least in 1940.

> Pac rules, p16, third paragraph of the Strat Bombing Rules
> "The
number of defending fighters that will intercept is decided by the owning player(s) after the attacker’s Combat Move phase
> is completed and before the Conduct Combat phase begins."

File: src/games/strategy/triplea/delegate/BattleTracker.java [hunk 1]
```diff
@@ -62,6 +65,7 @@
   // blitzed territories
   private final Set<Territory> m_blitzed = new HashSet<>();
   // territories where a battle occurred
+  // TODO: fix typo in name, 'fough' -> 'fought'
```
Comment (RoiEXLab) [inline src/games/strategy/triplea/delegate/BattleTracker.java]: This could be done in this PR

File: src/games/strategy/engine/data/Route.java [hunk 2]
```diff
@@ -355,8 +355,8 @@ public String toString() {
    *         the route consists of only a starting territory
    */
   public Territory getEnd() {
-    if (m_steps.size() == 0) {
-      return null;
+    if (m_steps.isEmpty()) {
+      return m_start;
```
Comment (ron-murhammer) [inline src/games/strategy/engine/data/Route.java]: Can you explain why this change is needed? Could this have adverse effects?

File: src/games/strategy/triplea/attachments/TerritoryAttachment.java [hunk 3]
```diff
@@ -138,15 +138,7 @@ public static Territory getFirstOwnedCapitalOrFirstUnownedCapital(final PlayerID
...
-    return get(t, false);
-  }
-
-  public static TerritoryAttachment get(final Territory t, final boolean allowNull) {
-    final TerritoryAttachment rVal = (TerritoryAttachment) t.getAttachment(Constants.TERRITORY_ATTACHMENT_NAME);
-    if (!allowNull && rVal == null && !t.isWater()) {
-      throw new IllegalStateException("No territory attachment for:" + t.getName());
-    }
-    return rVal;
+    return (TerritoryAttachment) t.getAttachment(Constants.TERRITORY_ATTACHMENT_NAME);
```
Comment (ron-murhammer) [inline src/games/strategy/triplea/attachments/TerritoryAttachment.java]: Doesn't this change functionality for null territories since we are passing false?

File: src/games/strategy/triplea/delegate/BattleTracker.java [hunk 4]
```diff
@@ -1084,6 +1100,30 @@ public void sendBattleRecordsToGameData(final IDelegateBridge aBridge) {
     }
   }
 
+
+  void fightStrategicBombingRuns(final IDelegateBridge delegateBridge) {
```
Comment (ron-murhammer) [inline src/games/strategy/triplea/delegate/BattleTracker.java]: Should this be 'fightStrategicBombingRaids' for consistency?

File: src/games/strategy/triplea/delegate/BattleTracker.java [hunk 5]
```diff
@@ -1084,6 +1100,30 @@ public void sendBattleRecordsToGameData(final IDelegateBridge aBridge) {
...
+    boolean bombing = true;
+    fightStrategicBombingRuns(delegateBridge, () -> getPendingBattleSites(bombing),
+            (territory, battleType) -> getPendingBattle(territory, bombing, battleType));
+  }
+
+  @VisibleForTesting
+  void fightStrategicBombingRuns(final IDelegateBridge delegateBridge, Supplier<Collection<Territory>> pendingBattleSiteSupplier,
```
Comment (ron-murhammer) [inline src/games/strategy/triplea/delegate/BattleTracker.java]: Should this be 'fightStrategicBombingRaids' for consistency?

File: src/games/strategy/triplea/delegate/BattleTracker.java [hunk 6]
```diff
@@ -1084,6 +1100,30 @@ public void sendBattleRecordsToGameData(final IDelegateBridge aBridge) {
...
+    // The rules say that bombing battles should be done first, so automatically do those combats.
+    for( final Territory t : pendingBattleSiteSupplier.get()) {
+      final IBattle airRaid = pendingBattleFunction.apply(t, BattleType.AIR_RAID);
+      if( airRaid != null ) {
+        airRaid.fight(delegateBridge);
+      }
+      final IBattle bombingRaid = pendingBattleFunction.apply(t, BattleType.BOMBING_RAID);
+      if( bombingRaid != null ) {
+        bombingRaid.fight(delegateBridge);
+      }
```
Comment (ron-murhammer) [inline src/games/strategy/triplea/delegate/BattleTracker.java]: What is the difference between an AIR_RAID and a BOMBING_RAID? Does the order of them matter?

Comment (simon33-2) [inline src/games/strategy/triplea/delegate/BattleTracker.java]: An Air Raid is the air battle preceding a bombing raid. In 1940 (optional in 1942.2) when one has a factory or base attacked, you have the option of sending up fighters to intercept the bombers. This defender decisions is (unfortunately) made after the attacker spawns the air raid. Theoretically, all air raid decisions should be made before any dice are rolled, although the consequences aren't too severe.

That's why I wanted to automate the air raids most of all. Group all the defender decisions together - scramble/kamikaze/intercept. It still won't be perfect because there is often an attacker decision about what facility will be attacked in 1940 (naval base/airbase/factory) so if there are two air raids in one turn there will still be a pass of control if the first one necessitates an attacker decision. However, it will resolve the 98% case of their only being one air raid.

Sorry if that is too long an answer.

Comment (ron-murhammer) [inline src/games/strategy/triplea/delegate/BattleTracker.java]: @simon33-2 I guess my confusion is I've never heard of it referred to as an "Air Raid" and I don't believe that term appears anywhere in the rules. I thought it was always called an "Air Battle" when the defender chose to use fighters to intercept.

Comment (simon33-2) [inline src/games/strategy/triplea/delegate/BattleTracker.java]: Fair enough. Air Raid and Bombing Raid are Triple-A terms. Not Axis and Allies terms.

File: src/games/strategy/triplea/attachments/TerritoryAttachment.java (hunk 3 above)
Comment (DanVanAtta) [inline src/games/strategy/triplea/attachments/TerritoryAttachment.java]: The test set up had a failure here. Looking at it, we are being overly restrictive. All usages of `get(final Territory t)` I could spot check for null return values. Though consider the case where we forgot a NPE check on return value and so we actually relied on the IllegalStateException. In that case we should still get a NPE since we're returning a null value, and the code will still error out. It's the same error behavior.

To be a bit more clear about what happened here though, we are putting the burden on the caller to require a non-null response. Note that we are actually in effect defaulting to 'true'. We are removing the condition controlled by `allowNull` altogether, effectively setting it to true (note the `!` that negates it, and when `allowNull` is true then we will never execute the if statement)

File: src/games/strategy/engine/data/Route.java (hunk 2 above)
Comment (DanVanAtta) [inline src/games/strategy/engine/data/Route.java]: It has never made much sense for `Route` to keep track of the head of the route differently from the rest of it. From other perspectives we can see there are problems with that configuration. There are a lot of details that clients have to know about the implementation in order to set the class up properly and use it. 

I hit an error here pretty quickly in test data while adding unit tests. It does make more sense for us to return the end of a route so long as it has one territory in it. The choice to return null if there is no end is pretty arbitrary. 

There are a number of usages of this method. There are 4 things that are giving me confidence this is okay to change:
- all unit tests pass
- manual testing in-game looks good
- AI vs AI game (mix of hard + easy) played through round 12 of a game without any issues
- inspection of the usages for a solid hour didn't raise any flags

For inspection we can cut out a number of classes. Non-pro AI's we do not care so much about, so Pro AI usages were checked out. They all seemed okay. MoveValidator had a lot of usages, but for the most part we know that is going to be okay with this type of change, and much of that could be skimmed. The other locations all seemed to be okay as well.

File: src/games/strategy/triplea/delegate/BattleTracker.java (hunk 1 above)
Comment (DanVanAtta) [inline src/games/strategy/triplea/delegate/BattleTracker.java]: Hate to say, but now that we are doing backwards compatible releases, we need to double check the blast radius of variable re-names when it comes to SaveGames. Anything that is a GameComponent must keep the same variable names for save games almost certainly. BattleTracker might be safe, not willing to check at the time. Also want to reduce code churn in case we choose to go ahead and fix the 'm_' naming as well.

**Comment by DanVanAtta** [comment]
re: method name of `fightStrategicBombingRuns`, I'm trying to avoid either calling it airRaid or SBR, since it captures both types of air battles. I went with `fightAirCombats` in the end. Happy to take suggestions if there is a better name. AIR_RAID was a surpirse to me as well, the enum value likely would be better named AIR_BATTLE. 

Please let me know if any feedback is inadequately addressed, otherwise I believe I've addressed all of the points so far raised.

File: src/games/strategy/engine/data/Route.java (hunk 2 above)
Comment (ron-murhammer) [inline src/games/strategy/engine/data/Route.java]: Alright. I'm ok moving forward with the change. I'd prefer changes like this in separate PRs or if in a PR like this then definitely commented since its pretty unclear why the change is being made.

File: src/games/strategy/triplea/attachments/TerritoryAttachment.java (hunk 3 above)
Comment (ron-murhammer) [inline src/games/strategy/triplea/attachments/TerritoryAttachment.java]: Right my concern is that the method above calls it with 'false' which is probably the method that is called publicly more often. That would make me think that we are relying on the exception being thrown in some cases. While you are correct that you'd most likely end up with NPE instead that would most likely be less descriptive of what the problem is.

File: src/games/strategy/triplea/delegate/BattleTracker.java [hunk 7]
```diff
@@ -1084,6 +1100,45 @@ public void sendBattleRecordsToGameData(final IDelegateBridge aBridge) {
     }
   }
 
+
+  /**
+   * 'Auto-fight' all of the air battles and strategic bombing runs.
+   * Auto fight means we automatically begin the fight without user action. This is to avoid clicks during the
+   * air battle and SBR phase, and to enforce game rules that these phases are fought first before any other combat.
+   */
+  void fightAirCombats(final IDelegateBridge delegateBridge) {
```
Comment (ron-murhammer) [inline src/games/strategy/triplea/delegate/BattleTracker.java]: So the problem with this name is there is also a battle type of AIR_BATTLE which you aren't fighting here. There are 3 different types of air related battle types: AIR_BATTLE, AIR_RAID, BOMBING_RAID. You are only fighting battles related to bombing here so I would suggest 'fightBombingBattles' or 'fightBombingRaids' or 'fightAirRaidsAndBombingRaids'. Personally I don't really like the battle type names we have but that is a bigger issue. Mostly just want to keep things consistent with the terms we already have.

File: src/games/strategy/triplea/delegate/BattleTracker.java (hunk 1 above)
Comment (ron-murhammer) [inline src/games/strategy/triplea/delegate/BattleTracker.java]: Essentially anything in data or delegate classes we need to be careful around.

File: src/games/strategy/engine/data/Route.java (hunk 2 above)
Comment (DanVanAtta) [inline src/games/strategy/engine/data/Route.java]: Change was required given the test case, a comment would have been good though, yeah.

File: src/games/strategy/triplea/attachments/TerritoryAttachment.java (hunk 3 above)
Comment (DanVanAtta) [inline src/games/strategy/triplea/attachments/TerritoryAttachment.java]: Perhaps, though I did not find any and am relatively sure they do not exist (or at least of any consequence). Yes we may lose some accuracy in the error message, though we can fill that detail in when we run across it. 
 I checked out usages of this function, there are not *that* many, we pretty consistently account for null possibly coming back. It really seems the throwing by default behavior is not correct/intended. It seems like this method was more used because it was there, rather than used with thought that we throw on no territory attachment found. Basically had the code been designed with the throwing behavior in mind, we wouldn't see all the null handling on the return of this method.
Bottom line and the point is that we gain in simplicity by not throwing, an easier test setup and just some less code, and it does not seem we are losing much of anything (and are certainly not introducing any new errors)

File: src/games/strategy/triplea/delegate/BattleTracker.java (hunk 7 above)
Comment (DanVanAtta) [inline src/games/strategy/triplea/delegate/BattleTracker.java]: Makes sense, 'fightBombingBattles' is a good suggestion, updated the method name to that.

**Comment by simon33-2** [comment]
Fight bombing battles still is confusing potentially because it seems to exclude the air raid step. Why not just fightRaids?

**Comment by DanVanAtta** [comment]
`raid` unqualified to me is a bit generic, I tend to think of land battles even. That's just me perhaps. 

`fightBombingBattles` seems reasonable enough, when not clear we should probably prefer longer and more descriptive names. There's been a lot of churn in the name, I think what we have now is really good enough.

**Comment by DanVanAtta** [comment]
On second consideration, @simon33-2 does have a good point. It's an easy follow-up/change, opened:
https://github.com/triplea-game/triplea/pull/1446 to rename to `fightAirRaidsAndStrategicBombing`

Getting really wordy with it! It's okay though compared to the extra clarity. Since the method call is pretty procedural, it's also particularly okay for the name to be verbose.

**Comment by DanVanAtta** [comment]
Thank you @simon33-2  for the work with #1426 , I hope you didn't mind too much me hijacking that and adding the test-case. You were right though to imply that 'testable' does not equate to 'easy to test'. It was pretty tricky to get a reasonable (isolated) test case fitted in.

**Comment by simon33-2** [comment]
If the change gets implemented, I'll be happy. I still want to auto-kill transports and undefended amphibious assaults (without other sea battles).
